content/posts/gitops-series-part-1.md

```diff
@@ -0,0 +1,616 @@
+       # Email address used for ACME registration
+       email: myemail@example.com
+       # Name of a secret used to store the ACME account private key
+       privateKeySecretRef:
+         name: letsencrypt-prod
+       # Enable the HTTP-01 challenge provider
+       http01: {}
+```
+
+The reason we have two of these is because Let's Encrypt has a rate-limiter on how often you can generate certificates. So while you are experimenting with things, it is safer to use the staging issuer. When things are all sorted, you can switch to the production-issuer.
+
+Go ahead and apply these two issuers with:
+
+```bash
+kubectl apply -f cert-manager/staging-issuer.yaml
+kubectl apply -f cert-manager/production-issuer.yaml
+```
+
+Now when we create an ingress for our applications, we can add some metadata to the ingress definition and the certificates will automatically be generated and stored in secrets.
+
+## Setting up CloudBees Core or CJD
+
+The final step in this flow is to either setup Core or CJD. We will add this portion, and more in the following posts.
+
+But before concluding this post, let's take a look at how we can put all of these steps into a Jenkinsfile.
+
+## Putting it all together.
+
+Since the whole objective here was to automate this process, it makes sense to use Jenkins to run the process. We can easily achieve GitOps by having every commit to master kick off our pipeline here.
+
+You'll note that we've added a couple of dependencies along the way that we'll need to make sure Jenkins will have access to. Thankfully, since we'll be running on Kubernetes, we can take advantage of the ephemeral, container-based agents. We can define a [pod template](https://jenkins.io/doc/pipeline/steps/kubernetes/) which will describe all of the containers we will need.
+
+In the root directory of my repo, I have created a `pod-template.yml` file:
+
+```yaml
+kind: Pod
+metadata:
+  name: gitops-pod
+spec:
+  containers:
+  - name: terraform
+    image: hashicorp/terraform:light
+    command:
+    - cat
+    tty: true
+    volumeMounts:
+      - name: gcp-credential
+        mountPath: /root/
+    env:
+      - name: GOOGLE_CLOUD_KEYFILE_JSON
+        value: "/root/gcp-service.json"
+      - name: GCP_SERVICE_ACCOUNT_FILE
+        value: "/root/gcp-service.json"
+      - name: GOOGLE_APPLICATION_CREDENTIALS
+        value: "/root/gcp-service.json"
+  - name: ansible
+    image: ldonleycb/ansible-ci:new
+    command:
+    - cat
+    tty: true
+    volumeMounts:
+      - name: gcp-credential
+        mountPath: /root/
+    env:
+      - name: GOOGLE_CLOUD_KEYFILE_JSON
+        value: "/root/gcp-service.json"
+      - name: GCP_SERVICE_ACCOUNT_FILE
+        value: "/root/gcp-service.json"
+      - name: GOOGLE_APPLICATION_CREDENTIALS
+        value: "/root/gcp-service.json"
+      - name: GCP_PROJECT
+        value: "my_project"
+      - name: GCP_CLUSTER_NAME
+        value: "my_cluster_name"
+  - name: kubectl
+    image: google/cloud-sdk:latest
```

Comment:
I would recommend using a specific version here as I believe it is a best practice - even a sha if you like.